README.md
## What is XServ-RPG?
XServ-RPG is an ambitious project aiming to facilitate writing detailed simulation games in a declarative way. 

## Setting expectations
Until further notice, set them low! This is a hobby project consisting of contributions from interested volunteers. 
The current main goal is to write enough of the framework to enable a complete demo game (unnamed interstellar alien bounty-hunting game). 

## Running Tests
Tests are run using pytest. To run tests, run `pip install pytest` in your venv and run `pytest .`

## Generating documentation
Documentation is generated with the pdoc package. Currently requires running `pdoc --html <package name>` for each package so this can be improved.


CHANGELOG.md
# Working
- Came up with basic component classes for `simulation` package, which will form the foundation of all simulations.
- Created complete worked-out example of gear-train implemented using `simulation` package, showing how a SimGraph can be used to compute the rotation of a gear mesh. 
- Added support for allowing gear meshes that "lock up" - i.e. are arranged in a way that makes turning impossible.

# Version 0.1
NOT RELEASED


test_gears.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-


## ToDo:
##   - Reduce line count somehow.


import pytest

from gears import Gear, GearTrain
# separate tests for readability
# arrange, act, assert

@pytest.fixture
def gear1():
    gear = Gear(10)
    gear.probability_of_breaking_per_tooth_click = 0 # NDH: having to move this into declaration feels like anti pattern
    return gear

@pytest.fixture
def gear2():
    gear = Gear(20)
    gear.probability_of_breaking_per_tooth_click = 0
    return gear

@pytest.fixture
def gear3():
    gear = Gear(10)
    gear.probability_of_breaking_per_tooth_click = 0
    return gear

@pytest.fixture
def gear_train(gear1, gear2, gear3):
    gtrain = GearTrain()
    gtrain.add_gear(gear1)
    gtrain.add_gear(gear2)
    gtrain.add_gear(gear3) # NDH: i think this reveals we should add a *args to the aggregate constructors or otherwise abstract GearTrains representation
    return gtrain

@pytest.fixture
def gear_train_three_gears_ring(gear1, gear2, gear3):
    gtrain = GearTrain()
    gtrain.add_gear(gear1)
    gtrain.add_gear(gear2)
    gtrain.add_gear(gear3)
    gtrain.add_gear_pair(gear3, gear1) ## close the loop
    return gtrain


def test_geartrain_implementation_in_simnode(gear1, gear2, gear3, gear_train):
    gear_train.teeth_moved = 5 ## Set input variable (probably not the cleanest way to do this - Dion)

    ## 2) Middle gear is broken, turning 5 teeth.
    gear1.angle = 0
    gear2.angle = 0
    gear3.angle = 0
    gear1.direction = 1
    gear2.direction = 1
    gear3.direction = 1
    gear1.teeth_moved = 5
    gear2.probability_of_breaking_per_tooth_click = 1

    ## save state variables for comparison
    gear2_state_before = gear2.get_state_variables()
    gear2_input_before = gear2.get_input_variables()
    gear3_state_before = gear3.get_state_variables()
    gear3_input_before = gear3.get_input_variables()

    assert gear_train.update()
    # NDH: I see in the implementation update this returns False if unsuccessful.
    # It's more pythonic to define an exception "UpdateException" that can be thrown from those classes.
    # That way we can catch it where we need to without doing boolean compares. This is especially useful when we start working with try catch game loop.

    assert (gear1.angle, gear1.direction, gear1.broken) == (180, 1, False)
    assert (gear2.angle, gear2.direction, gear2.broken) == (0,  -1, True)
    assert (gear3.angle, gear3.direction, gear3.broken) == (0,   1, False) ## gear should not break if turned by 0 teeth.

    ## causal chain broken by gear2, gear3 should be unaffected
    assert gear2.get_state_variables() != gear2_state_before ## gear2 broke, but
    assert gear2.get_input_variables() != gear2_input_before ## gear2 update was attempted
    assert gear3.get_state_variables() == gear3_state_before
    assert gear3.get_input_variables() == gear3_input_before
    ## notice direction switched even though gear2 did not rotate, since direction is an input variable rather than a state variable

def test_gears_turn_then_update_angle_and_direction(gear1, gear2, gear3, gear_train): #NDH: code duplication in test signature means I should probably group these
    ## 1) All gears OK, turning 5 teeth.
    gear_train.teeth_moved = 5
    assert gear_train.update()
    assert (gear1.angle, gear1.direction, gear1.broken) == (180,  1, False)
    assert (gear2.angle, gear2.direction, gear2.broken) == (-90, -1, False)
    assert (gear3.angle, gear3.direction, gear3.broken) == (180,  1, False)

    
def test_odd_numbered_circular_train_locks_up(gear1, gear2, gear3, gear_train_three_gears_ring):
    gear_train = gear_train_three_gears_ring
    print(gear_train.gears)
    print(gear_train.sim_graph.arrows)
    gear_train.teeth_moved = 5
    gear1_state_prior = gear1.get_state_variables()
    gear2_state_prior = gear2.get_state_variables()
    gear3_state_prior = gear3.get_state_variables()
    print(gear1_state_prior)
    print(gear2_state_prior)
    print(gear3_state_prior)
    assert not gear_train.update() ## we expect this to fail because odd-numbered circular gear trains lock up
    assert gear1_state_prior == gear1.get_state_variables()
    assert gear2_state_prior == gear2.get_state_variables()
    assert gear3_state_prior == gear3.get_state_variables()
    


test_simulation.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-


import pytest


from simulation import SimNode, SimArrow, SimGraph


## this could be factored out into a non-test utilities class - will leave that for when it becomes necessary.
class BooleanNode(SimNode):
    """
    Simple switch with on/off state only.
    """
    def __init__(self):
        self.input_state = False

    def update(self):
        """
        Switches are in a sense simply just inputs. They have no internal mechanism, so there's nothing to do here.
        """
        return True

    def get_state_variables(self) -> dict:
        return dict()

    def get_input_variables(self) -> dict:
        return {"input_state": self.input_state}


class BooleanArrow(SimArrow):


    def __init__(self, source:BooleanNode, target:BooleanNode):
        self.source = source
        self.target = target

    def effect(self) -> bool:
        """
        Simply copy the input state to the output node.
        """
        self.target.input_state = self.source.input_state
        return True


@pytest.fixture
def boolean_node_1():
    return BooleanNode()

@pytest.fixture
def boolean_node_2():
    return BooleanNode()

@pytest.fixture
def boolean_arrow_1(boolean_node_1, boolean_node_2):
    return BooleanArrow(boolean_node_1, boolean_node_2)


def test_change_propagation(boolean_arrow_1, boolean_node_1, boolean_node_2):
    """
    Check that arrows actually propagate input changes effectually.
    """
    assert False == boolean_node_1.input_state
    assert False == boolean_node_2.input_state
    boolean_node_1.input_state = True
    assert boolean_arrow_1.update() ## propagate change
    assert True == boolean_node_1.input_state
    assert True == boolean_node_2.input_state


source/test_gears.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-


import pytest

from gears import Gear, GearTrain
# separate tests for readability
# arrange, act, assert

@pytest.fixture
def gear1():
    gear = Gear(10)
    gear.probability_of_breaking_per_tooth_click = 0 # NDH: having to move this into declaration feels like anti pattern
    return gear

@pytest.fixture
def gear2():
    gear = Gear(20)
    gear.probability_of_breaking_per_tooth_click = 0
    return gear

@pytest.fixture
def gear3():
    gear = Gear(10)
    gear.probability_of_breaking_per_tooth_click = 0
    return gear

@pytest.fixture
def gear_train(gear1, gear2, gear3):
    gtrain = GearTrain()
    gtrain.add_gear(gear1)
    gtrain.add_gear(gear2)
    gtrain.add_gear(gear3) # NDH: i think this reveals we should add a *args to the aggregate constructors or otherwise abstract GearTrains representation
    return gtrain

@pytest.fixture
def gear_train_three_gears_ring(gear1, gear2, gear3):
    gtrain = GearTrain()
    gtrain.add_gear(gear1)
    gtrain.add_gear(gear2)
    gtrain.add_gear(gear3)
    gtrain.add_gear_pair(gear3, gear1) ## close the loop
    return gtrain


def test_geartrain_implementation_in_simnode(gear1, gear2, gear3, gear_train):
    gear_train.teeth_moved = 5 ## Set input variable (probably not the cleanest way to do this - Dion)

    ## 2) Middle gear is broken, turning 5 teeth.
    gear1.angle = 0
    gear2.angle = 0
    gear3.angle = 0
    gear1.direction = 1
    gear2.direction = 1
    gear3.direction = 1
    gear1.teeth_moved = 5
    gear2.probability_of_breaking_per_tooth_click = 1

    ## save state variables for comparison
    gear2_state_before = gear2.get_state_variables()
    gear2_input_before = gear2.get_input_variables()
    gear3_state_before = gear3.get_state_variables()
    gear3_input_before = gear3.get_input_variables()

    assert gear_train.update()
    # NDH: I see in the implementation update this returns False if unsuccessful.
    # It's more pythonic to define an exception "UpdateException" that can be thrown from those classes.
    # That way we can catch it where we need to without doing boolean compares. This is especially useful when we start working with try catch game loop.

    assert (gear1.angle, gear1.direction, gear1.broken) == (180, 1, False)
    assert (gear2.angle, gear2.direction, gear2.broken) == (0,  -1, True)
    assert (gear3.angle, gear3.direction, gear3.broken) == (0,   1, False) ## gear should not break if turned by 0 teeth.

    ## causal chain broken by gear2, gear3 should be unaffected
    assert gear2.get_state_variables() != gear2_state_before ## gear2 broke, but
    assert gear2.get_input_variables() != gear2_input_before ## gear2 update was attempted
    assert gear3.get_state_variables() == gear3_state_before
    assert gear3.get_input_variables() == gear3_input_before
    ## notice direction switched even though gear2 did not rotate, since direction is an input variable rather than a state variable

def test_gears_turn_then_update_angle_and_direction(gear1, gear2, gear3, gear_train): #NDH: code duplication in test signature means I should probably group these
    ## 1) All gears OK, turning 5 teeth.
    gear_train.teeth_moved = 5
    assert gear_train.update()
    assert (gear1.angle, gear1.direction, gear1.broken) == (180,  1, False)
    assert (gear2.angle, gear2.direction, gear2.broken) == (-90, -1, False)
    assert (gear3.angle, gear3.direction, gear3.broken) == (180,  1, False)

    
def test_odd_numbered_circular_train_locks_up(gear1, gear2, gear3, gear_train_three_gears_ring):
    gear_train = gear_train_three_gears_ring
    gear_train.teeth_moved = 5
    gear1_state_prior = gear1.get_state_variables()
    gear2_state_prior = gear2.get_state_variables()
    gear3_state_prior = gear3.get_state_variables()
    assert not gear_train.update() ## we expect this to fail because odd-numbered circular gear trains lock up
    assert gear1_state_prior == gear1.get_state_variables()
    assert gear2_state_prior == gear2.get_state_variables()
    assert gear3_state_prior == gear3.get_state_variables()
    


roguelike.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-


"""
XServ Roguelike Version 0.1
Authors:
    - Dion Bridger
"""


import random


from abc import ABC as AbstractBaseClass
from abc import abstractmethod


class SimNode(AbstractBaseClass):
    
    """
    SimNode instances wrap sets of simulation state variables.
    They are the most fine-grained components out of which complex simulations are assembled.
    """
    
    @abstractmethod
    def update(self) -> bool:
        """
        When a SimNode's update() method is called, some of the variables belonging to the node can be updated to reflect changes made to its
        input variables.
        """
        pass


class SimArrow(AbstractBaseClass):
    
    """
    Arrows represent unidirection causal relationships between nodes.
    All arrows must implement an 'effect' which copies values from the source to the target.
    """
    
    source: SimNode
    target: SimNode

    def update(self) -> bool:
        if self.effect():
            return self.target.update()
        return False

    @abstractmethod
    def effect(self) -> bool:
        """
        The effect method should implement the causal relationship represented by the arrow. This means copying the value of at least one of the variables
        of the source node into a variable in the target node, but not vice versa.
        """
        pass


class SimGraph:

    """
    A SimGraph is composed of a set of connected SimArrow instances, and may include cycles. If the graph includes cycles then the causal chain will stop
    once chain of atom updates comes back around to the first atom updated.
    """

    ## Status: stub class
    ## ToDo: 
    ##   - cycle checking
    ##   - conflict / deadlock checking (consider three gears all connected, none should be able to actually turn)
    ##   - "intelligent" arrow updates, don't update down a dead causal chain

    def __init__(self):
        self.arrows = list()

    def add_arrow(self, arrow:SimArrow):
        self.arrows.append(arrow)

    def update(self) -> bool:
        for a in self.arrows:
            a.update()


  ## ############## ##
 ## ## Examples ## ##
## ############## ##

# ToDo:
#   - move examples into a separate file.
#   - think through how to model causal graphs with closed loops. E.g. gears arranged in a ring.

class Gear(SimNode):
    
    def __init__(self, teeth):
        
        ## static parameters
        self.teeth = teeth
        
        ## variables
        self.angle = 0
        self.teeth_moved = 0
        self.broken = False
        self.probability_of_breaking_per_tooth_click = 1/100
    
    def update(self) -> bool:
        if self.broken:
            return False
        teeth_moved = 0
        for i in range(self.teeth_moved):
            self.broken = random.random() < self.probability_of_breaking_per_tooth_click 
            if self.broken:
                break
            teeth_moved += 1
        self.teeth_moved = teeth_moved
        self.angle = (360*teeth_moved/self.teeth)%360
        return True


class GearPair(SimArrow):

    def __init__(self, source:Gear, target:Gear):
        self.source = source
        self.target = target

    def effect(self) -> bool:
        if not self.source.broken:
            self.target.teeth_moved = self.source.teeth_moved
            return True
        return False


class GearTrain(SimNode):
    
    def __init__(self, start_gear:Gear):
        self.teeth_moved = 0
        self.gears = [start_gear,]
        self.arrows = []
    
    def add_gear(self, gear):
        self.arrows.append(GearPair(self.gears[-1], gear))
        self.gears.append(gear)
    
    def update(self):
        self.gears[0].update()
        for a in self.arrows:
            if not a.update():
                break


def gear_example():
    g0 = Gear(10)
    g1 = Gear(20)
    g2 = Gear(10)
    gtrain = GearTrain(g0)
    gtrain.add_gear(g1)
    gtrain.add_gear(g2)
    g0.teeth_moved = 5
    gtrain.update()
    print("1) all gears OK, turning 5 teeth")
    print(g0.angle, g0.broken)
    print(g1.angle, g1.broken)
    print(g2.angle, g2.broken)

    g0.angle = 0
    g1.angle = 0
    g2.angle = 0
    g0.teeth_moved = 5
    g1.broken = True
    gtrain.update()
    print("\n2) middle gear is broken, turning 5 teeth")
    print(g0.angle, g0.broken)
    print(g1.angle, g1.broken)
    print(g2.angle, g2.broken)
    

if __name__ == "__main__":
    main()


source/simulation.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-


"""
XServ Roguelike Version 0.1
Authors:
    - Dion Bridger
"""


import random


from abc import ABC as AbstractBaseClass
from abc import abstractmethod

from collections import defaultdict



class GraphUpdateException(Exception):
    """
    Represents an /unexpected/ failure during an update operation of a SimGraph.
    """
    pass



class SimNode(AbstractBaseClass):

    """
    SimNode instances wrap sets of simulation state variables.
    They are the nodeic components out of which complex simulations are assembled.
    """

    @abstractmethod
    def update(self) -> bool:
        """
        When a SimNode's update() method is called, some of the variables belonging to the node can be updated to reflect changes made to its
        input variables.
        """
        pass

    @abstractmethod
    def get_state_variables(self) -> dict:
        """
        Get a dictionary containing all the state variables of the SimNode. This can be used to check whether the state of a node was altered without
        having to check the individual member variables of the node, or even having to know what they actually are.
        Setting values in dictionary returned should NOT change the value of the SimNode.
        """
        pass

    @abstractmethod
    def get_input_variables(self) -> dict:
        """
        Return a subset of self.__dict__ containing all variables considered to be 'input' variables.
        """
        pass

    def get_variables(self) -> dict:
        """
        Return all simulation variables held by this instance.
        """
        return {**self.get_input_variables(), **self.get_state_variables()}


class SimArrow(AbstractBaseClass):

    """
    Arrows represent unidirection causal relationships between nodes.
    All arrows must implement an 'effect' which copies values from the source to the target.
    """

    source: SimNode
    target: SimNode

    def update(self) -> bool:
        """
        Effect an update of the target node's state given the state of the input node.
        """
        if self.effect():
            return self.target.update()
        return False

    @abstractmethod
    def effect(self) -> bool:
        """
        The effect method should implement the causal relationship represented by the arrow. This means copying the value of at least one of the variables
        of the source node into a variable in the target node, but not vice versa.
        """
        pass


class SimGraph:

    """
    A SimGraph is composed of a set of connected SimArrow instances, and may include cycles. If the graph includes cycles then the causal chain will stop
    once chain of atom updates comes back around to the first atom updated.
    """

    def __init__(self):
        self.starting_arrow = None
        self.arrows = []
        self.arrow_to_arrows = defaultdict(list)

    def add_arrow(self, arrow:SimArrow) -> None:
        """
        Extend the graph with a new causal arrow.
        """
        if None==self.starting_arrow:
            self.starting_arrow = arrow
        for a in self.arrows:
            if a.target == arrow.source:
                self.arrow_to_arrows[a].append(arrow)
            if arrow.target == a.source:
                self.arrow_to_arrows[arrow].append(a)
        self.arrows.append(arrow)

    def update(self) -> bool:
        """
        Compute the next state of the graph given its current state. If we return False then no state variables of any SimNode in the graph should
        change after we return. If we return True then all active causal arrows should have effected their changes before we exit.
        """
        arrow_stack = [self.starting_arrow,]
        nodes_to_prior_states = {self.starting_arrow.source: self.starting_arrow.source.get_variables()}
        self.starting_arrow.source.update()
        success = True ## Using a flag rather than an exception as failure to update state is not necessarily an error.
        while arrow_stack:
            ## select target node to update
            arrow = arrow_stack.pop()
            source = arrow.source
            target = arrow.target
            target_prior_state = nodes_to_prior_states.get(target)
            if not target_prior_state:
                nodes_to_prior_states[target] = target.get_variables()
            ## Perform update 
            chain_done = not arrow.update()
            if target_prior_state:
                success &= (target_prior_state == target.get_variables())
                chain_done = True
            else:
                arrow_stack.extend(self.arrow_to_arrows[arrow])
            ## Undo all state changes if an inconsistency obtains
            if not success:
                ## Undo all state updates
                for node, prior_state in nodes_to_prior_states.items():
                    node.__dict__.update(prior_state)
            ## Check for premature exit
            if chain_done:
                break
        return success


gears.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import random

from simulation import SimNode, SimGraph, SimArrow

class Gear(SimNode):

    instance_number = 0

    def __init__(self, teeth):
        self.instance_number = Gear.instance_number
        Gear.instance_number += 1

        ## static parameters
        self.teeth = teeth

        ## variables
        self.angle = 0 ## should be in the range [-180,180]
        self.teeth_moved = 0
        self.direction = 1 ## 1=clockwise. -1=anticlockwise
        self.broken = False
        self.probability_of_breaking_per_tooth_click = 0 ## default 0 so as not to mess with unit tests

    def __repr__(self):
        return f"Gear[{self.instance_number}]"

    def update(self) -> bool:
        if self.broken:
            return False
        teeth_moved = 0
        for i in range(self.teeth_moved):
            self.broken = random.random() < self.probability_of_breaking_per_tooth_click
            if self.broken:
                break
            teeth_moved += 1
        self.teeth_moved = teeth_moved
        angle_delta = self.direction*(360*teeth_moved/self.teeth)
        angle = (self.angle+angle_delta)%360
        if 180 < angle:
            angle += -360
        elif angle < -180:
            angle += 360
        self.angle = angle
        return True

    def get_state_variables(self) -> dict:
        return {
            "angle": self.angle,
            "broken": self.broken,
            "probability_of_breaking_per_tooth_click": self.probability_of_breaking_per_tooth_click
        }

    def get_input_variables(self) -> dict:
        return {
            "direction": self.direction,
            "teeth_moved": self.teeth_moved,
        }


class GearPair(SimArrow):

    def __init__(self, source:Gear, target:Gear):
        self.source = source
        self.target = target

    def __repr__(self):
        return f"GearPair[{self.source.instance_number}, {self.target.instance_number}]"

    def effect(self) -> bool:
        if not self.source.broken:
            self.target.direction = -1*self.source.direction
            self.target.teeth_moved = self.source.teeth_moved
            return True
        return False


class GearTrain:

    def __init__(self):
        self.gears = list()
        self.teeth_moved = 0
        self.sim_graph = SimGraph()

    def _append_gear(self, gear):
        if gear not in self.gears:
            self.gears.append(gear)

    def add_gear(self, gear):
        self._append_gear(gear)
        if 2<=len(self.gears):
            gp = GearPair(self.gears[-2], self.gears[-1])
            self.sim_graph.add_arrow(gp)

    def add_gear_pair(self, source_gear, target_gear):
        self._append_gear(source_gear)
        self._append_gear(target_gear)
        gp = GearPair(source_gear, target_gear)
        self.sim_graph.add_arrow(gp)

    def update(self):
        self.gears[0].teeth_moved = self.teeth_moved
        return self.sim_graph.update()
